### ISSUE TITLE
Using Material.Scheme.top apparently breaks the debugger in Elm 0.18

### ISSUE THREAD

**Issue by frankschmitt** [debois/elm-mdl#270]
Running this simple Elm Mdl program (a stripped-down version of the Counter example):

`module Main exposing (..)

import Html exposing (..)
import Html.Attributes exposing (href, class, style)
import Material
import Material.Scheme
import Material.Button as Button
import Material.Options as Options exposing (css, onClick)
import Material.Layout as Layout

-- MODEL
type alias Mdl = Material.Model
type Msg = Mdl (Material.Msg Msg)

type alias Model = Material.Model

model : Model
model = Material.model

-- UPDATE
update : Msg -> Model -> ( Model, Cmd Msg )
update msg model = model ! []

-- VIEW
view : Model -> Html Msg
view model =
    Layout.render Mdl
        model
        [ Layout.fixedHeader
        ]
        { header = [ h1 [ style [ ( "padding", "2rem" ) ] ] [ text "Counter" ] ]
        , drawer = []
        , tabs = ( [], [] )
        , main = [ viewBody model ]
        }

-- body
viewBody : Model -> Html Msg
viewBody model =
    div
        [ ] 
        [ text "Hello from Mdl" 
        ]
        -- |> Material.Scheme.top

-- main program
main : Program Never Model Msg
main =
    program
        { init = ( model, Cmd.none )
        , view = view
        , subscriptions = always Sub.none
        , update = update
        }
`

works perfectly fine with Elm's debugger (v 0.18, run via elm-reactor), but if you un-comment this line:
`      -- |> Material.Scheme.top`
  
the **Explore history** button in Elm's debugger isn't working anymore (at first glance, it seems there's no event handler for it).

**Comment by debois** [comment]
Huh. Interesting. @vipentti do you have time to take a look?

**Comment by debois** [comment]
@frankschmitt: Could you speculate what's causing `Scheme.top` to break the debugger? We might want to submit an issue against reactor, since I don't think we're doing anything untoward.

**Comment by frankschmitt** [comment]
@debois (I'm just getting started with Web development, so please take this with a grain of salt)
The event handler for the click event is attached in both cases (with Scheme.top and without), so this looks ok. I noticed that  elm-mdl-layout-main has a z-index of 1; everything related to the Elm debugger seems to have a z-index of auto, so this might be the problem (I don't know how click events are handled for overlapping elements with different z-indices - does the first one consume the event and the rest never gets to "see" it? Or does it "bubble" all the way through to the last one?)

**Comment by frankschmitt** [comment]
I did some further experimenting - re-attaching the event handlers in Firebug's console with

/* assign on-click handler by CSS class - 
   see http://stackoverflow.com/a/3808886/610979 */
function assignOnClickToCSSClass(matchClass, fn) {
    var elems = document.getElementsByTagName('*'), i;
    for (i in elems) {
        if((' ' + elems[i].className + ' ').indexOf(' ' + matchClass + ' ')
                > -1) {
            elems[i].onclick = fn;
        }
    }
}
assignOnClickToCSSClass('elm-mini-controls-import-export', 
  function() { console.log('ELM mini controls import export'); }
);
assignOnClickToCSSClass('mdl-layout__container',
  function() { console.log('MDL'); }
);

got me these results:
- with Material.Scheme.top, the event handler for mdl-layout-container is triggered when I click on the Elm Debugger import/export button
- without Materia.Scheme.top, the event handler for elm-mini-controls-import-export is triggered when I click on the Elm Debugger import/export button
- manually setting the z-index for the Elm debugger with

.elm-mini-controls-import-export {
  z-index: 100;
}

and injecting this file into my app by modifiying Main.elm:

viewBody model =
    div
        [ ] 
        [ text "Hello from Mdl" 
        , Html.node "link" [ Html.Attributes.rel "stylesheet", Html.Attributes.href "style.css" ] []
        ]
        |> Material.Scheme.top

apparently doesn't help, either - I see that the z-index is indeed 100, but still, the wrong event handler is triggered.

I'm somewhat out of my depth - should I try to do some further digging / provide more information? (I'd like to avoid cluttering Github with unnecessary / unhelpful information).

**Comment by debois** [comment]
This is great—thanks for taking the trouble to investigate this!

Could you use the inspector in your browser to try and figure out which computed CSS properties are different in the two cases? E.g., using `Scheme.top` versus loading Google CSS in the page header?

**Comment by frankschmitt** [comment]
I did some further experimenting:
- instead of using Material.Scheme.top, directly referencing the style sheet via         

, Html.node "link" [ Html.Attributes.rel "stylesheet", Html.Attributes.href "https://code.getmdl.io/1.3.0/material.min.css" ] []

in Elm

Result: same as with using Material.Scheme.top - the event for the Elm mini controls doesn't work 
 (what I forgot to mention before: the cursor doesn't change for the Elm mini controls when using Material Design - without it, the cursor changes to a pointer - apparently because elm-mini-controls has

cursor: pointer;

in its CSS)
- when I manually replace

.mdl-layout__container {
    height: 100%;
}

with

.mdl-layout__container {
    height: 80%;
}

the Elm mini controls work correctly. As far as I can tell, with 100%, the MDL layout container completely "overlaps" the Elm mini controls and "swallows" the events (not sure if I'm using the correct terminology here)

So apparently, the problem isn't specific to Material.Scheme.top, but with Material Design per se.

**Comment by dosarf** [comment]
Another ad-hoc workaround: if I set the z-index: 2 for the CSS class elm-layout (temporarily in the browser itself) then both the MDL stuff and the Elm Debugger worked.

**Comment by damukles** [comment]
I agree with @frankschmitt and @dosarf It is not only appearing with Material.Scheme.top, it happens from time to time to me, but I never use Material.Scheme.top. In case it happens, overwriting the z-index of the elm-overlay to 1 or more helps to access the debugger/history. But still, a click on import or export says:
```
The Types.Msg type of your program cannot be reliably serialized for history files.
Functions cannot be serialized, nor can values that contain functions. This is a problem in these places:
Material.Dispatch.Config can contain functions and JSON decoders.
```
Don't know if that is expected.

**Comment by wfranzini** [comment]
Hi,

I'm having trouble with this issue: what's the best way to deal with it?

thanks
walter

**Comment by martindlarsson** [comment]
I think I am having the same or related issue. I'm using both top och footer. When the debugger is activated it shows up as expected but I cannot interact with it. The footer sometimes partially covers the debugger and I guess it as something to do with z-index and that the footer prevents click events on the debugger.

**Comment by joelclermont** [comment]
A quick workaround: drop this in a css file
`.elm-overlay {
  z-index: 999;
}`

Now you can see history and click through messages, but import/export won't work as long as Material.Dispatch.Config contains the decoder and function.

**Comment by tad-lispy** [comment]
I think this is more of an issue with Elm Debugger than Elm MDL. Any application with elements having element with style like this:

```css
element {
    position: absolute;
    width: 100%;
    height: 100%;
    z-index: 1;
}
```

will cover `.elm-mini-controls` and make it *unclickable*.

Two solutions come to mind to make sure `.elm-mini-controls` are on top of application:

- Setting high `z-index` 
  
  easy but crappy - application style may and often do set arbitrarily high z-indices

- ~~Or mount fullscreen Elm application inside `div.elm-overlay` before `div.elm-mini-controls`, so the DOM tree would look something like that:~~

  ```html
  <div class="elm-overlay">
    <div class="mdl-layout__container"><!-- or any other Elm application DOM Tree--></div>
    <style><!-- ... --></style>
    <div class="elm-mini-controls"><!-- ... --></div>
  </div>
  ```
  
  **edit: This is a wrong approach - it breaks scrolling.**
